### PR TITLE
Feature: Control number of vmapped envs in evaluator using `arch.num_envs` 

### PR DESCRIPTION
**Pull request by OmaymaMahjoub**
## What?
Modify the evaluator to limit the number of vmapped envs to `arch.num_envs` when the total number of evaluation episodes `arch.num_eval_episodes` exceeds this limit (instead of parallelise all the `arch.num_eval_episodes`). In such cases, evaluations are conducted in sequential batches, with each batch containing `arch.num_envs` parallel envs.

## Why?
Limiting parallel evaluations to num_envs prevents out-of-memory issues by avoiding vmap over all episodes at once.

### CODE REVIEW

File: mava/evaluator.py [hunk 1]
```diff
@@ -104,41 +104,68 @@ def not_done(carry: Tuple) -> bool:
...
 
+    def evaluator_fn(trained_params: FrozenDict, key: chex.PRNGKey) -> ExperimentOutput[EvalState]:
+        """Evaluator function"""
+        n_devices = len(jax.devices())
+
+        # Calculate episodes to evaluate per device.
+        episodes_per_device = config.arch.num_eval_episodes * eval_multiplier // n_devices
+        # Determine parallel evaluation batch size.
+        # Limited by architecture's max environments or episodes per device.
+        parallel_eval_batch_size = min(config.arch.num_envs, episodes_per_device)
```
Comment (sash-a) [inline mava/evaluator.py]: Parallel can be a bit confusing, I thought at first this was the `pmap` dim. Please also change below
```suggestion
        # Determine how many eval episodes to vmap over.
        # Limited by architecture's max environments or episodes per device.
        num_vmapped_episodes = min(config.arch.num_envs, episodes_per_device)
```

File: mava/evaluator.py [hunk 2]
```diff
@@ -104,41 +104,68 @@ def not_done(carry: Tuple) -> bool:
...
-        """Evaluator function."""
+    def parallel_evaluator(
+        carry: Tuple[FrozenDict, chex.PRNGKey], _: Any
+    ) -> Tuple[Tuple[FrozenDict, chex.PRNGKey], ExperimentOutput[EvalState]]:
+        """Evaluates the agent on multiple environments in parallel."""
+        trained_params, key = carry
 
-        # Initialise environment states and timesteps.
         n_devices = len(jax.devices())
+        episodes_per_device = config.arch.num_eval_episodes * eval_multiplier // n_devices
```
Comment (WiemKhlifi) [inline mava/evaluator.py:114]: Sadly jax doesn't allow us to carry these fixed values (`parallel_eval_batch_size`) 😢

**Review by WiemKhlifi — DISMISSED** [review]
LGTM, thanks @OmaymaMahjoub for making things more flexible 🙏

File: mava/evaluator.py [hunk 3]
```diff
@@ -104,41 +104,68 @@ def not_done(carry: Tuple) -> bool:
...
+        n_devices = len(jax.devices())
+
+        # Calculate episodes to evaluate per device.
+        episodes_per_device = config.arch.num_eval_episodes * eval_multiplier // n_devices
+        # Determine parallel evaluation batch size.
+        # Limited by architecture's max environments or episodes per device.
+        parallel_eval_batch_size = min(config.arch.num_envs, episodes_per_device)
+        # Compute the number of sequential evaluation batches required per device
+        # to cover all episodes.
+        sequential_eval_batches = episodes_per_device // parallel_eval_batch_size
```
Comment (sash-a) [inline mava/evaluator.py]: If you make keys of shape `(sequential_eval_batches, num_vmapped_episodes)` then you don't need the extra method, you can just call:

```python
jax.lax.scan(jax.vmap(eval_one_episode), None, eval_states)
```

I like this for two reasons, you don't have to calculate `parallel_eval_batch_size` twice and it is also clear that you're scanning and then vmapping over episodes. But this is a big change so if you don't have time it's ok. I think the evaluator is due for a big overhaul anyways

**Review by sash-a — CHANGES_REQUESTED** [review]
Two small things, one would take a while though so if there's no time we can wait till we refactor the evaluator

File: mava/evaluator.py (hunk 3 above)
Comment (WiemKhlifi) [reply mava/evaluator.py]: I back this idea also 🔥

File: mava/evaluator.py (hunk 2 above)
Comment (OmaymaMahjoub) [reply mava/evaluator.py:114]: I am quite sure there is a better way to implement this, but we can keep it as it is at the moment and create an issue calling for cleaning and readibility of the evaluator

File: mava/evaluator.py (hunk 3 above)
Comment (OmaymaMahjoub) [reply mava/evaluator.py]: I see so even the eval_init will be created inside the one episode, I will give it a try

**Comment by OmaymaMahjoub** [comment]
This PR has been archived because PR #1084 has resolved the issue and includes additional enhancements.
